Linter definitions and linter config refactor (#235)
trunk-io/docs: check/configuration/custom-linters/reference.md

* define the linters (`linters`),
* list linters to enable and disable (`enabled` and `disabled`)
* define file categories (`files`)
* list required `runtimes`, `downloads`, `environments` ,
* and additional cross-linter settings

### `files`

`files`: Definitions of filetypes

Every linter must define the set of filetypes it applies to in the `lint.files` section.

New filetypes are defined with the name and extensions properties.
They may also include the comments properties to describe what style of
comments are used in these files.

This is how the C++ source filetype is defined.

```yaml
lint:
  files:
    - name: c++-source
      extensions:
        - C
        - cc
        - cpp
        - cxx
      comments:
        - slashes-block
        - slashes-inline
```

### `runtimes`

`runtimes`: Node, python, cargo, etc. Used to define or override a runtime environment for package management. [See Runtimes](../../advanced-setup/runtimes.md) 

[//]: # (`linters`: maps a linter name to it's corresponding linter definition)

### `definitions`

`definitions`: Where you define or override linter settings. See [Linter Definition Config](#lint-config-definitions)

### `enabled`

> `enabled`: list of enabled linters

Author: I don't see anything about the extended map form. Can you give me an example?

global linter settings, see [Lint Config](reference.md#lint-config)

### `name`

`name` Is the name of the linter. This property will be used to refer to the linter in other 
parts of the config, for example, in the list of enabled linters.

[//]: # (`type` is the type of output the linter produces. **link to output types**)

[//]: # (`command`, string, command to run: ex: `[$workspace/tools/linters/clang-format, --assume-filename=${path}]`)

[//]: # (`success_codes` is the list of success codes that indicates linter generated data and should be processed. Some linters return 0 for nothing to lint and 1 when there are issues.)

[//]: # (`error_codes` is the list of error codes this linter will return when it hit an internal failure and couldn't generate results.)

### `direct_configs`

__If you have one of these config files, you're definitely using this linter__

### `affects_cache`

`affects_cache`: the list of files that affect the cache results of this linter.

### `good_without_config`

`good_without_config`: *optional boolean*. Indicates whether this linter is recommended without 
the user tuning its configuration. Prefer `suggest_if`.

### `files`

> `files` is a list of file types listed in the `lint.files` section that this linter applies to. 

Reviewer: Delete and move this text to the commands section files. The files description there is incorrect.

### `runtime`

`runtime`: RuntimeType, Which runtime, if any, to require to be setup for this linter.

### `package`

`package`: string, What primary package to install, if using a package manager runtime

### `extra_packages`

`extra_packages`: list of strings, Extra packages to install, versions are optional

### `download`

`download`: *string*. The download url. You must provide either runtime + packages or download, not both. Using runtimes is preferred. [See Runtimes](../../advanced-setup/runtimes.md)

`issue_url_format`: string, a format string that accepts issue codes for links to issues docs.

### `run_from_root_target`

`run_from_root_target`: *string*. Walk up to find this file to detect the run from directory. 
Prefer `run_from` at the command level.

### `hold_the_line`

`hold_the_line`: *optional boolean*. Whether hold-the-line will be done for this linter or not.

### `read_output_from`

`read_output_from`,  Tell the parser where to expect output from for reading (stdout, stderr, 
tmp file). [See Output Sources](output-types.md#output-sources)

### `prepare_command`

`prepare_command`, ex. `[tflint, --init]`

[//]: # (`version`, **???** )

### `known_good_version`

> `known_good_version`: The version `trunk init` inits with. If not present Trunk will query the 

Reviewer: This is not quite correct. This field will only be used for tools where trunk does not know how to automatically query the latest version. Currently, trunk can query the latest version for all package managers and downloads hosted on github.

version from its runtime or use the default version from the download.

### `version_command`

`version_command`:  Version check commands

### `enabled`

`enabled`: *optional boolean*. Whether this linter is enabled.

### `batch`

`batch`: *optional boolean*. Combine multiple files into the same execution.

### `run_when`

> `run_when`, Should this linter be run only in CI? (or locally, vscode, monitor)

Reviewer: Possible values are: cli, monitor, ci, lsp

### `in_place`

> `in_place`: *optional boolean*. Applies to formatters - does the formatter rewrite the file as 

Reviewer: We should only say that this tells trunk that a formatting command will result in the file being rewritten in place.

### `run_timeout`

`run_timeout`: *duration string*. Describes how long a linter can run before timing out. 
[See timeouts](../readme.md#timeout)

### `disabled`

`disabled`: *optional boolean*: Whether linter is actively disabled (and will not be recommended)
and will not run (overrides enabled).

### `commands`
`commands`: Commands exposed by this linter. [See Linter Command Definition](#linter-command-definition)

### `deprecated`

`deprecated`: *string*. Deprecated information (set when linter is deprecated)

### `known_bad_versions`

`known_bad_versions`: *string list*. Versions of a linter that are known to be broken 
and should not be run with Trunk. We will fall back to a `known_good_version` 
if init or upgrade chooses something in this set.

### `plugin_url`

`plugin_url`: *string*, a plugin url for reporting issues.

### `query_compile_commands`

`query_compile_commands`, *optional boolean*

### `idempotent`

`idempotent`: *optional boolean*. Indicates whether a linter is idempotent w.r.t. config + 
source code inputs.  e.g. semgrep fetches rules from the internet, so it's not idempotent . If 

> set, causes cache_ttl to default to 24h

Reviewer: ```suggestion
set, will only cache results a duration of `cache_ttl`.
```

### `cache_ttl`

> `cache_ttl`, duration string, How long cached results are kept before they expire

Reviewer: ```suggestion
`cache_ttl`, duration string, If this linter is not idempotent, this is how long cached results are kept before they expire. Defaults to 24hrs.
```

### `target`

> `target`: optional string, what target does this run on

Reviewer: ```suggestion
`target`: optional string, what target does this run on. Defaults to `${file}`.
```

We should expand on other examples here.

Author: I added three code examples.

### `suggest_if`

How to determine if this linter should be auto-enabled/recommended. Possible 
values are `never`, `config_present`, and `files_present`.

* `config_present` will auto-enable a linter if Trunk sees any [`direct_config`](#direct_configs) for it .
* `files_present` will auto-enable a linter if Trunk sees any file type that it operates on.
* `never` will never auto-enable this linter.

Trunk curates the values of `suggest_if` for all linters in the [plugins](https://github.com/trunk-io/plugins) repo.

### `supported_platforms`

Platform constraint. If incompatible, renders a notice.

### `tools`

`tools`, string list

### `main_tool`

> `main_tool`, *string*. The linter's main tool. This should be the tool that provides the binary 
> for the linter. It will generally be the same as the linter's name, but may differ if the linter 
> is a wrapper. The main tool is used to determine the linter's version.

Reviewer: ```suggestion
`main_tool`, *string*. If your linter depends on more than a single tool, and none of the tools has the same name as the linter, then you will need to specify which is the main tool here. It will be used to version the tool from the linter's enabled version.
```

### `path_format`

`path_format`, Whether to use the platform-specific paths or generic "/". Default native.

## Linter Command Definition

The linter command definitions are defined in `lint.linter.commands`. A single linter can have 
multiple commands if it is used in different ways.  

*Note:*. If you define the executable to run here (the command definition), then you should *not*
define it also in the linter definition. Defining it here as a command is preferred.

### `name`

`name`: *string*. A unique name for this command (some tools expose multiple commands, format, 
lint, analyze, etc.).

### `output`

`output`: which type of output this linter produces. [See Output Types](output-types.md).

### `prepare_run`

`prepare_run`: an extra command to run before running a linter.

### `run`

> `run`: The command to run a linter. 

Reviewer: Do we have a table of all possible template expansions for `run` somewhere?

Author: Do you mean the variables that are exposed?  We have this list. https://docs.trunk.io/check/configuration/custom-linters#command

Author: I added an example of how dart uses the `${target}` variable, and also linked to the list of all variables.

> [//]: # (TODO: link to examples in main docs)

Reviewer: TODO.

### `files`

`files` File configs for this command. If absent, will use the whole linter's `file_configs`. Note: this defines the entire list of file configs for this command, rather than appending to those of the entire lint command.

> [//]: # (TODO: link to example)

Reviewer: TODO.

Author: Should `files` be specified at the linter or command level? Looking through our plugins repo, it seems to usually be at the linter level. ex: prettier

### `run_when`

`run_when`: Should this command be run only in CI? One of `cli`, `lsp`, `monitor`, or `ci`.

### `success_codes`

`sucess_codes:` List of exit codes that indicates linter ran successfully. **Unrelated to 
whether or not there were issues reported by the linter**

### `error_codes`

`error_codes`: List of exit codes this linter will return when it hit an internal failure and 
couldn't generate results.  **A linter should set either success codes or error codes, but not both.**

### `no_issues_codes`

`no_issues_codes`: List of exit codes that Trunk will use to assume there were no issues without 
parsing the output.

### `read_output_from`

`read_output_from`: Tell parser where to expect output from for reading.
[See Output Sources](output-types.md#output-sources)

> [//]: # (TODO: **check values** &#40;stdout, stderr, tmp_file&#41; )

Reviewer: TODO.

> [//]: # (TODO: **link to output docs**)

Reviewer: TODO.

### `run_from`

`run_from`: What current working directory to run the linter from. 

> [//]: # (TODO: **link to section with more  examples**.)

Reviewer: TODO.

### `batch`

> `batch`: *optional boolean*. Combine multiple files into the same execution.

Reviewer: ```suggestion
`batch`: *optional boolean*. Combine multiple files into the same execution. if true, the `${target}` template substitution in the `run` field may expand into multiple files.
```

### `in_place`

> `in_place`: *optional boolean*. Applies to formatters - does the formatter rewrite the file as 

Reviewer: Looks at comments from Linter dedinition.

opposed to reading its contents from STDIN and outputting formatted contents to STDOUT?

### `sandbox_type`

> `sandbox_type`: How to create the sandbox if it needs to be created.

Reviewer: Delete. Not ready to publically document this yet.

### `allow_empty_files`

`allow_empty_files`: *optional boolean*. Skip linting empty files for this linter. Trunk will 
assume there are no linters if the file is empty.

### `std_in`

`std_in`: *optional boolean*. Should the command be fed the file on standard input?

### `cache_results`

`cache_results`: *optional boolean*. Indicates if this linter wants to cache results. 

[//]: # (TODO: **explain how trunk does caching in another page**)

### `disable_upstream`

`disable_upstream`: *optional boolean*,  Whether this linter supports comparing against the 
upstream version of this file.  

> [//]: # (TODO: **link to a deeper explanation for when you set this to false**)

Reviewer: TODO.

### `enabled`

`enabled`: *optional boolean*. Whether the command is enabled to run when the linter is run. 
Allows some commands of a linter to be run by default without others.

### `formatter`

`formatter`: *optional boolean*. Whether this command is a formatter and should be included in 
`trunk fmt`.

### `parser`

`parser`: The definition of a parser that will transform the output of the linter into SARIF. 
Not needed if linter is already output SARIF. [See Output Types](./output-types.md)

### `parse_regex`

`parse_regex`: *string*. A regular expression used to support regex parsing.
[See Regex output type](./output-types.md#regex)

### `maximum_file_size`

`maximum_file_size`: *number*. The maximum file size in bytes for input files to the linter.

### `idempotent`

> `idempotent`: *optional boolean*. Indicates whether a linter is idempotent with config 

Reviewer: Look at comments from linter defintion. Also this should only exist here.TODO.

and source code inputs. For example, `semgrep` fetches rules from the internet, so it is not 
idempotent.  If set, it causes `cache_ttl` to default to `24h`. This affects how long linter 
outputs are cached.

### `cache_ttl`

> `cache_ttl`: *string*. How long cached results are kept before they expire.  Only applies if `idempotent: false`

Reviewer: Look at comments from linter defintion. Also this should only exist here.TODO.

> [//]: # (TODO: link to deeper explanation** )

Reviewer: TODO.

### `version`

`version`: *optional string*, Version constraint.  

> [//]: # (TODO: link to deeper explanation of version constraints with example)

Reviewer: TODO.

### `target`

`target`, *optional string*, What target does this run on. By default, the target is the modified 
source code file. Some linters operate on a whole repo or directory. 

> [//]: # (TODO: link to deeper docs.)

Reviewer: TODO.
